### PR TITLE
[FIRRTL] Canonicalize multibit_mux with narrow index

### PR DESCRIPTION
**Pull request by uenoku**
This adds a canonicalization to optimize multibit_mux whose index has narrow width. Close https://github.com/llvm/circt/issues/7361

### CODE REVIEW

File: lib/Dialect/FIRRTL/FIRRTLFolds.cpp [hunk 1]
```diff
@@ -1656,6 +1656,17 @@ LogicalResult MultibitMuxOp::canonicalize(MultibitMuxOp op,
     return success();
   }
 
+  // If the index width is narrower than the size of inputs, drop front
+  // elements.
+  auto indexWidth = op.getIndex().getType().getBitWidthOrSentinel();
+  uint64_t inputSize = op.getInputs().size();
+  if (indexWidth < 64 && 1ull << indexWidth < inputSize) {
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/FIRRTLFolds.cpp:1663]: Is `1ull` portable across windows and Linux?

**Review by seldridge — APPROVED** [review]
LGTM

While, this looks right to me, it is worthwhile to verify (locally test) that the behavior is doing to right thing for edge cases, e.g., zero width.

File: lib/Dialect/FIRRTL/FIRRTLFolds.cpp (hunk 1 above)
Comment (uenoku) [reply lib/Dialect/FIRRTL/FIRRTLFolds.cpp:1663]: I think it's fine as there are several existing use cases in LLVM and CIRCT repo.

Comment (dtzSiFive) [reply lib/Dialect/FIRRTL/FIRRTLFolds.cpp:1663]: long long is at least 64bits since c++11, according to https://en.cppreference.com/w/cpp/language/types .

That said (only looking at the code added here):

indexWidth is signed (int32_t), and in particular can be negative (!).  Shift by a negative amount is UB! Please add a check for that, as even without UB it's logically not what we want either.

**Review by dtzSiFive — COMMENTED** [review]
Great work identifying this and putting together the canonicalizer!

On my "TODO" is to bolster our knownbits-like capabilities, and with that hat on I can't help but wonder about generalizing this beyond index being narrow directly.  Future work, but since you're pushing on this just thought I'd mention it 👍 .

Anyway, generally looks good but there's an important issue that needs to be addressed before landing this, see my comment.  Thanks!

**Comment by uenoku** [comment]
> On my "TODO" is to bolster our knownbits-like capabilities, and with that hat on I can't help but wonder about generalizing this beyond index being narrow directly. Future work, but since you're pushing on this just thought I'd mention it 👍 .

Yeah that sounds great idea. I think naively computing knownbits could be expensive (as shown in https://github.com/llvm/circt/blob/e8033f1928ca78d6fa35cdf9cc8bb809b99b5422/lib/Dialect/Comb/CombFolds.cpp#L618-L627) so I guess a dedicated pass would be needed. Though computing known bits is exactly bit-sensitive IMCP so it's worth extending the lattice in IMCP.

**Review by dtzSiFive — CHANGES_REQUESTED** [review]
Friendly upgrade to "request changes" since this is reachable UB.  I don't have the details paged in but the expected change (dropping widths, I think?) to one of the tests was all it took 👍 .

Just making sure this gets addressed, thanks!

**Comment by uenoku** [comment]
Thanks! I added a condition and test to ensure that the width is not negative number

**Review by dtzSiFive — APPROVED** [review]
LGTM! Thanks!

(sorry for delay, missed the update 👍 )

**Comment by uenoku** [comment]
Np, thank you for review!
